File: mlflow/tracing/trace_manager.py
import contextlib
import logging
import threading
from dataclasses import dataclass, field
from typing import Generator, Sequence

from mlflow.entities import LiveSpan, Trace, TraceData, TraceInfo
from mlflow.entities.model_registry import PromptVersion
from mlflow.environment_variables import MLFLOW_TRACE_TIMEOUT_SECONDS
from mlflow.prompt.constants import LINKED_PROMPTS_TAG_KEY
from mlflow.tracing.utils.prompt import update_linked_prompts_tag
from mlflow.tracing.utils.timeout import get_trace_cache_with_timeout
from mlflow.tracing.utils.truncation import set_request_response_preview

_logger = logging.getLogger(__name__)


# Internal representation to keep the state of a trace.
# Dict[str, Span] is used instead of TraceData to allow access by span_id.
@dataclass
class _Trace:
    info: TraceInfo
    span_dict: dict[str, LiveSpan] = field(default_factory=dict)
    prompts: list[PromptVersion] = field(default_factory=list)

    def to_mlflow_trace(self) -> Trace:
        trace_data = TraceData()
        for span in self.span_dict.values():
            # Convert LiveSpan, mutable objects, into immutable Span objects before persisting.
            trace_data.spans.append(span.to_immutable_span())

        set_request_response_preview(self.info, trace_data)
        return Trace(self.info, trace_data)

    def get_root_span(self) -> LiveSpan | None:
        for span in self.span_dict.values():
            if span.parent_id is None:
                return span
        return None


@dataclass
class ManagerTrace:
    """
    Wrapper around a trace and its associated prompts.
    """

    trace: Trace
    prompts: Sequence[PromptVersion]


class InMemoryTraceManager:
    """
    Manage spans and traces created by the tracing system in memory.
    """

    _instance_lock = threading.RLock()
    _instance = None

    @classmethod
    def get_instance(cls):
        if cls._instance is None:
            with cls._instance_lock:
                if cls._instance is None:
                    cls._instance = InMemoryTraceManager()
        return cls._instance

    def __init__(self):
        # In-memory cache to store trace_od -> _Trace mapping.
        self._traces = get_trace_cache_with_timeout()

        # Store mapping between OpenTelemetry trace ID and MLflow trace ID
        self._otel_id_to_mlflow_trace_id: dict[int, str] = {}
        self._lock = threading.RLock()  # Lock for _traces

    def register_trace(self, otel_trace_id: int, trace_info: TraceInfo):
        """
        Register a new trace info object to the in-memory trace registry.

        Args:
            otel_trace_id: The OpenTelemetry trace ID for the new trace.
            trace_info: The trace info object to be stored.
        """
        # Check for a new timeout setting whenever a new trace is created.
        self._check_timeout_update()
        with self._lock:
            self._traces[trace_info.trace_id] = _Trace(trace_info)
            self._otel_id_to_mlflow_trace_id[otel_trace_id] = trace_info.trace_id

    def register_span(self, span: LiveSpan):
        """
        Store the given span in the in-memory trace data.

        Args:
            span: The span to be stored.
        """
        if not isinstance(span, LiveSpan):
            _logger.debug(f"Invalid span object {type(span)} is passed. Skipping.")
            return

        with self._lock:
            trace_data_dict = self._traces[span.request_id].span_dict
            trace_data_dict[span.span_id] = span

    def register_prompt(self, trace_id: str, prompt: PromptVersion):
        """
        Register a prompt to link to the trace with the given trace ID.

        Args:
            trace_id: The ID of the trace to which the prompt belongs.
            prompt: The prompt version to be registered.
        """
        with self._lock:
            if prompt not in self._traces[trace_id].prompts:
                self._traces[trace_id].prompts.append(prompt)

            # NB: Set prompt URIs in trace tags for linking. This is a short-term solution until
            # LinkPromptsToTraces endpoint is implemented in the backend.
            # TODO: Remove this once LinkPromptsToTraces endpoint is implemented in the backend.
            try:
                current_tag = self._traces[trace_id].info.tags.get(LINKED_PROMPTS_TAG_KEY)
                updated_tag = update_linked_prompts_tag(current_tag, [prompt])
                self._traces[trace_id].info.tags[LINKED_PROMPTS_TAG_KEY] = updated_tag
            except Exception:
                _logger.debug(f"Failed to update prompts tag for trace {trace_id}", exc_info=True)
                raise

    @contextlib.contextmanager
    def get_trace(self, trace_id: str) -> Generator[_Trace | None, None, None]:
        """
        Yield the trace info for the given trace ID..
        This is designed to be used as a context manager to ensure the trace info is accessed
        with the lock held.
        """
        with self._lock:
            yield self._traces.get(trace_id)

    def get_span_from_id(self, trace_id: str, span_id: str) -> LiveSpan | None:
        """
        Get a span object for the given trace_id and span_id.
        """
        with self._lock:
            trace = self._traces.get(trace_id)

        return trace.span_dict.get(span_id) if trace else None

    def get_root_span_id(self, trace_id) -> str | None:
        """
        Get the root span ID for the given trace ID.
        """
        with self._lock:
            trace = self._traces.get(trace_id)

        if trace:
            for span in trace.span_dict.values():
                if span.parent_id is None:
                    return span.span_id

        return None

    def get_mlflow_trace_id_from_otel_id(self, otel_trace_id: int) -> str | None:
        """
        Get the MLflow trace ID for the given OpenTelemetry trace ID.
        """
        return self._otel_id_to_mlflow_trace_id.get(otel_trace_id)

    def set_trace_metadata(self, trace_id: str, key: str, value: str):
        """
        Set the trace metadata for the given request ID.
        """
        with self.get_trace(trace_id) as trace:
            if trace:
                trace.info.trace_metadata[key] = value

    def pop_trace(self, otel_trace_id: int) -> ManagerTrace | None:
        """
        Pop trace data for the given OpenTelemetry trace ID and
        return it as a ManagerTrace wrapper containing the trace and prompts.
        """
        with self._lock:
            mlflow_trace_id = self._otel_id_to_mlflow_trace_id.pop(otel_trace_id, None)
            internal_trace = self._traces.pop(mlflow_trace_id, None) if mlflow_trace_id else None
            if internal_trace is None:
                return None
            return ManagerTrace(
                trace=internal_trace.to_mlflow_trace(), prompts=internal_trace.prompts
            )

    def _check_timeout_update(self):
        """
        TTL/Timeout may be updated by users after initial cache creation. This method checks
        for the update and create a new cache instance with the updated timeout.
        """
        new_timeout = MLFLOW_TRACE_TIMEOUT_SECONDS.get()
        if new_timeout != getattr(self._traces, "timeout", None):
            if len(self._traces) > 0:
                _logger.warning(
                    f"The timeout of the trace buffer has been updated to {new_timeout} seconds. "
                    "This operation discards all in-progress traces at the moment. Please make "
                    "sure to update the timeout when there are no in-progress traces."
                )

            with self._lock:
                # We need to check here again in case this method runs in parallel
                if new_timeout != getattr(self._traces, "timeout", None):
                    self._traces = get_trace_cache_with_timeout()

    @classmethod
    def reset(cls):
        """Clear all the aggregated trace data. This should only be used for testing."""
        if cls._instance:
            with cls._instance._lock:
                cls._instance._traces.clear()
            cls._instance = None

File: libs/typescript/core/src/core/trace_manager.ts
import { LiveSpan, Span } from './entities/span';
import { TraceInfo } from './entities/trace_info';
import { Trace } from './entities/trace';
import { TraceData } from './entities/trace_data';
import {
  REQUEST_RESPONSE_PREVIEW_MAX_LENGTH,
  SpanAttributeKey,
  TraceMetadataKey,
} from './constants';

/**
 * Internal representation to keep the state of a trace.
 * Uses a Map<string, LiveSpan> instead of TraceData to allow access by span_id.
 */
class _Trace {
  info: TraceInfo;
  spanDict: Map<string, LiveSpan>;

  constructor(info: TraceInfo) {
    this.info = info;
    this.spanDict = new Map<string, LiveSpan>();
  }

  /**
   * Convert the internal trace representation to an MLflow Trace object
   */
  toMlflowTrace(): Trace {
    // Convert LiveSpan, mutable objects, into immutable Span objects before persisting
    const traceData = new TraceData([...this.spanDict.values()] as Span[]);

    const root_span = traceData.spans.find((span) => span.parentId == null);
    if (root_span) {
      // Only set previews if they haven't been explicitly set by updateCurrentTrace
      if (!this.info.requestPreview) {
        this.info.requestPreview = getTruncatedPreview(
          root_span._span.attributes[SpanAttributeKey.INPUTS] as string,
          'user',
        );
      }
      if (!this.info.responsePreview) {
        this.info.responsePreview = getTruncatedPreview(
          root_span._span.attributes[SpanAttributeKey.OUTPUTS] as string,
          'assistant',
        );
      }

      // TODO: Remove this once the new trace table UI is available that is based on MLflow V3 trace.
      // Until then, these two metadata are still used to render the "request" and "response" columns.
      this.info.traceMetadata[TraceMetadataKey.INPUTS] = this.info.requestPreview;
      this.info.traceMetadata[TraceMetadataKey.OUTPUTS] = this.info.responsePreview;
    }

    return new Trace(this.info, traceData);
  }
}

/**
 * Generate a truncated preview string from span inputs/outputs.
 * Extracts message text content from known chat formats (OpenAI messages,
 * choices, Responses API) before truncating, so the preview shows readable
 * text instead of cut-off JSON.
 *
 * Mirrors the Python SDK's `_get_truncated_preview` in
 * `mlflow/tracing/utils/truncation.py`.
 */
function getTruncatedPreview(inputsOrOutputs: string, role: string): string {
  if (!inputsOrOutputs) {
    return '';
  }

  const maxLength = REQUEST_RESPONSE_PREVIEW_MAX_LENGTH;
  let content: string | null = null;

  try {
    const obj = JSON.parse(inputsOrOutputs);
    if (obj && typeof obj === 'object') {
      const messages = tryExtractMessages(obj as Record<string, unknown>);
      if (messages && messages.length > 0) {
        const msg = getLastMessageByRole(messages, role);
        content = getTextContentFromMessage(msg);
      }
    }
  } catch {
    // Not valid JSON — use as-is
  }

  const preview = content || inputsOrOutputs;
  if (preview.length <= maxLength) {
    return preview;
  }
  return preview.slice(0, maxLength - 3) + '...';
}

function tryExtractMessages(obj: Record<string, unknown>): Record<string, unknown>[] | null {
  // OpenAI ChatCompletion request format: { messages: [...] }
  if (Array.isArray(obj.messages)) {
    return obj.messages.filter(isMessage);
  }

  // OpenAI ChatCompletion response format: { choices: [{ message: {...} }] }
  if (
    Array.isArray(obj.choices) &&
    obj.choices.length > 0 &&
    typeof obj.choices[0] === 'object' &&
    obj.choices[0] != null
  ) {
    const msg = (obj.choices[0] as Record<string, unknown>).message;
    if (isMessage(msg)) {
      return [msg];
    }
  }

  // OpenAI Responses API request format: { input: [...] }
  if (Array.isArray(obj.input)) {
    return obj.input.filter(isMessage);
  }

  // OpenAI Responses API response format: { output: [...] }
  if (Array.isArray(obj.output)) {
    return obj.output.filter(isMessage);
  }

  // ResponsesAgent input: { request: { ... } }
  if (obj.request && typeof obj.request === 'object') {
    return tryExtractMessages(obj.request as Record<string, unknown>);
  }

  return null;
}

function isMessage(item: unknown): item is Record<string, unknown> {
  return (
    item != null &&
    typeof item === 'object' &&
    'role' in (item as Record<string, unknown>) &&
    'content' in (item as Record<string, unknown>)
  );
}

function getLastMessageByRole(
  messages: Record<string, unknown>[],
  role: string,
): Record<string, unknown> {
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i].role === role) {
      return messages[i];
    }
  }
  return messages[messages.length - 1];
}

function getTextContentFromMessage(message: Record<string, unknown>): string | null {
  const content = message.content;
  if (typeof content === 'string') {
    return content;
  }
  if (Array.isArray(content)) {
    for (const part of content) {
      if (typeof part === 'string') {
        return part;
      }
      if (
        part &&
        typeof part === 'object' &&
        'type' in (part as Record<string, unknown>) &&
        ((part as Record<string, unknown>).type === 'text' ||
          (part as Record<string, unknown>).type === 'output_text')
      ) {
        return (part as Record<string, unknown>).text as string;
      }
    }
  }
  return null;
}

/**
 * Manage spans and traces created by the tracing system in memory.
 * This class is implemented as a singleton.
 */
export class InMemoryTraceManager {
  private static _instance: InMemoryTraceManager | undefined;

  // In-memory cache to store trace_id -> _Trace mapping
  // TODO: Add TTL to the trace buffer similarly to Python SDK
  private _traces: Map<string, _Trace>;
  // Store mapping between OpenTelemetry trace ID and MLflow trace ID
  private _otelIdToMlflowTraceId: Map<string, string>;

  // Store the last active trace ID
  lastActiveTraceId: string | undefined;

  /**
   * Singleton pattern implementation
   */
  static getInstance(): InMemoryTraceManager {
    if (InMemoryTraceManager._instance == null) {
      InMemoryTraceManager._instance = new InMemoryTraceManager();
    }
    return InMemoryTraceManager._instance;
  }

  private constructor() {
    this._traces = new Map<string, _Trace>();
    this._otelIdToMlflowTraceId = new Map<string, string>();
  }

  /**
   * Register a new trace info object to the in-memory trace registry.
   * @param otelTraceId The OpenTelemetry trace ID for the new trace
   * @param traceInfo The trace info object to be stored
   */
  registerTrace(otelTraceId: string, traceInfo: TraceInfo): void {
    this._traces.set(traceInfo.traceId, new _Trace(traceInfo));
    this._otelIdToMlflowTraceId.set(otelTraceId, traceInfo.traceId);
  }

  /**
   * Store the given span in the in-memory trace data.
   * @param span The span to be stored
   */
  registerSpan(span: LiveSpan): void {
    const trace = this._traces.get(span.traceId);
    if (trace) {
      trace.spanDict.set(span.spanId, span);
    } else {
      console.debug(`Tried to register span ${span.spanId} for trace ${span.traceId}
                     but trace not found. Please make sure to register the trace first.`);
    }
  }

  /**
   * Get the trace for the given trace ID.
   * Returns the trace object or null if not found.
   * @param traceId The trace ID to look up
   */
  getTrace(traceId: string): _Trace | null {
    return this._traces.get(traceId) || null;
  }

  /**
   * Get the MLflow trace ID for the given OpenTelemetry trace ID.
   * @param otelTraceId The OpenTelemetry trace ID
   */
  getMlflowTraceIdFromOtelId(otelTraceId: string): string | null {
    return this._otelIdToMlflowTraceId.get(otelTraceId) || null;
  }

  /**
   * Get the span for the given trace ID and span ID.
   * @param traceId The trace ID
   * @param spanId The span ID
   */
  getSpan(traceId?: string | null, spanId?: string | null): LiveSpan | null {
    if (traceId == null || spanId == null) {
      return null;
    }
    return this._traces.get(traceId)?.spanDict.get(spanId) || null;
  }

  /**
   * Pop trace data for the given OpenTelemetry trace ID and return it as
   * a ready-to-publish Trace object.
   * @param otelTraceId The OpenTelemetry trace ID
   */
  popTrace(otelTraceId: string): Trace | null {
    const mlflowTraceId = this._otelIdToMlflowTraceId.get(otelTraceId);
    if (!mlflowTraceId) {
      console.debug(`Tried to pop trace ${otelTraceId} but no trace found.`);
      return null;
    }

    this._otelIdToMlflowTraceId.delete(otelTraceId);
    const trace = this._traces.get(mlflowTraceId);
    if (trace) {
      this._traces.delete(mlflowTraceId);
      return trace.toMlflowTrace();
    }
    console.debug(`Tried to pop trace ${otelTraceId} but trace not found.`);
    return null;
  }

  /**
   * Clear all the aggregated trace data. This should only be used for testing.
   */
  static reset(): void {
    if (InMemoryTraceManager._instance) {
      InMemoryTraceManager._instance._traces.clear();
      InMemoryTraceManager._instance._otelIdToMlflowTraceId.clear();
      InMemoryTraceManager._instance = undefined;
    }
  }
}
